### PR TITLE
Add route /api/dev/start-build for use by Slack slash commands

### PR DESCRIPTION
**Pull request by islemaster**
Introduces an API we can hit with Slack slash commands, like `/restart-staging-build` or `/restart-test-build`, to trigger rebuilds in our CI environments without needing to SSH to the respective machines.

I've already set up the [`/restart-staging-build`](https://codedotorg.slack.com/services/116523429974) slash command and added its token to our staging Chef configuration, so this should start working for staging as soon as it's merged.

I'll comment on particular decisions and link to Slack documentation inline.

### CODE REVIEW

File: pegasus/routes/dev_routes.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+# Used to restart builds on staging/test via Slack /start-build slash command.
+post '/api/dev/start-build' do
+  # Forbidden in production because it's a dev route
+  # Forbidden in development because it won't work anyway
+  forbidden! if [:development, :production].include? rack_env
+
+  dont_cache
+
+  forbidden! unless params[:token] == CDO.slack_start_build_token
```
Comment (islemaster) [inline pegasus/routes/dev_routes.rb:11]: Non-200 responses result in a user-facing error ([docs](https://api.slack.com/slash-commands#responding_to_a_command)).  That's fine - we want to be clear that this doesn't work in certain environments, and not without a valid token.

File: pegasus/routes/dev_routes.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  # Don't create build-started if it already exists; let the requester know.
+  if File.file?(BUILD_STARTED_PATH)
+    return "I can't do that #{params[:user_name]} - a build is already queued"
```
Comment (islemaster) [inline pegasus/routes/dev_routes.rb:15]: This plaintext response will go back to the requesting user, but not to anyone else in the room ([docs](https://api.slack.com/slash-commands#responding_to_a_command)).  That seems appropriate for this particular situation, where the user will get what they want but their request is redundant with another.

File: pegasus/routes/dev_routes.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+  # Create the build-started file so a new build will start within one minute
+  system "touch #{BUILD_STARTED_PATH}"
+
+  # Notify the room
+  content_type :json
+  JSON.pretty_generate({
+    text: "#{rack_env.to_s.capitalize} build restarted by #{params[:user_name]}",
+    response_type: 'in_channel'
+  })
```
Comment (islemaster) [inline pegasus/routes/dev_routes.rb:26]: Here we use a JSON response to add the `in_channel` response type, causing the user's command and the affirmative response to be visible to everyone in their channel.

**Comment by islemaster** [comment]
Autocomplete for the new command: 
image

**Comment by islemaster** [comment]
The response to the new command right now, since this change hasn't been merged and there's nothing at the route it's hitting:
image

File: pegasus/test/test_dev_routes.rb [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        })
+        assert_equal 403, pegasus.last_response.status
+      end
+    end
+
+    it 'generates a start_build file if none exists' do
+      in_rack_env(:test) do
+        pegasus = make_test_pegasus
+        begin
+          refute File.file? BUILD_STARTED_PATH
```
Comment (aoby) [inline pegasus/test/test_dev_routes.rb:99]: redundant with line 19 in `before`

File: pegasus/test/test_dev_routes.rb [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+          assert_equal original_modify_time, File.mtime(BUILD_STARTED_PATH)
+
+          # Check response to requester
+          assert_equal 200, pegasus.last_response.status
+          assert_equal "I can't do that Dave - a build is already queued", pegasus.last_response.body
+        ensure
+          system "rm -f #{BUILD_STARTED_PATH}"
+        end
+      end
+    end
```
Comment (aoby) [inline pegasus/test/test_dev_routes.rb:139]: nice testing!

File: pegasus/test/test_dev_routes.rb (hunk 4 above)
Comment (islemaster) [reply pegasus/test/test_dev_routes.rb:99]: Intentional - I felt like it communicated the before/after assertion of this particular test case better.

**Pull request merged**